RULES/hdl/questa.go

```diff
@@ -2,75 +2,486 @@ package hdl
+func CompileIp(ctx core.Context, rule Simulation, ip Ip, 
+	             deps []core.Path, incs []core.Path) ([]core.Path, []core.Path) {
+	for _, sub_ip := range ip.Ips() {
+		deps, incs = CompileIp(ctx, rule, sub_ip, deps, incs)
+	}
+	deps, incs = CompileSrcs(ctx, rule, deps, incs, ip.Sources())
+
+	return deps, incs
+}
+
+// Compile compiles the IP dependencies and source files of a simulation rule.
+func Compile(ctx core.Context, rule Simulation) []core.Path {
+	incs := []core.Path{}
+	deps := []core.Path{}
+
+	for _, ip := range rule.Ips {
+		deps, incs = CompileIp(ctx, rule, ip, deps, incs)
+	}
+	CompileSrcs(ctx, rule, deps, incs, rule.Srcs)
+
+	return deps
+}
+
+// Optimize creates and optimized version of the design optionally including
+// coverage recording functionality. The optimized design unit can then conveniently
+// be simulated using 'vsim'.
+func Optimize(ctx core.Context, rule Simulation, deps []core.Path) {
+	top := "board"
+	if rule.Top != "" {
+		top = rule.Top
+	}
+	
+	cover_flag := ""
+	log_file_suffix := "/vopt.log"
+	if Coverage.Value() {
+		cover_flag = "+cover"
+		log_file_suffix = "/vopt_cover.log"
+	}
+
+	log_files := []core.Path{}
+	targets := []string{}
+	params := []string{}
+	if rule.Params != nil {
+		for key, _ := range rule.Params {
+			log_files = append(log_files, rule.Path().WithSuffix(key))
+			targets = append(targets, rule.Target() + key)
+			params = append(params, key)
+		}
+	} else {
+		log_files = append(log_files, rule.Path())
+		targets = append(targets, rule.Target())
+		params = append(params, "")
+	}
+
+	for i := range log_files {
+		log_file := log_files[i].WithSuffix(log_file_suffix)
+		target := targets[i]
+		param_set := params[i]
+
+		// Skip if we already have a rule
+		if rules[log_file.String()] {
+			return
+		}
+
+		cmd := fmt.Sprintf("vopt %s %s +acc=%s -l %s -work %s %s -o %s", 
+											common_flags, cover_flag, Access.Value(),
+											log_file.String(), rule.Lib(), top, target)
+
+		// Set up parameters
+		if param_set != "" {
+			// Check that the parameters exist
+			if params, ok := rule.Params[param_set]; ok {
+				// Add parameters for all generics
+				for param, value := range params {
+					cmd = fmt.Sprintf("%s -G %s=%s", cmd, param, value)
+				}
+			} else {
+				log.Fatal(fmt.Sprintf("parameter set '%s' not defined for Simulation target '%s'!", 
+				          params, rule.Name))
+			}
+		}
+		
+		// Add the rule to run 'vopt'.
+		ctx.AddBuildStep(core.BuildStep{
+			Out:   log_file,
+			Ins:   deps,
+			Cmd:   cmd,
+			Descr: fmt.Sprintf("vopt: %s %s", rule.Lib() + "." + top, target),
+		})
+
+		// Note that we created this rule
+		rules[log_file.String()] = true
+	}
 }
 
-type SimulationQuesta struct {
-	Name    string
-	Srcs    []core.Path
-	Ips     []Ip
-	Libs    []string
-	Verbose bool
+// BuildQuesta will compile and optimize the source and IPs associated with the given
+// rule.
+func BuildQuesta(ctx core.Context, rule Simulation) {
```

Comment:
Is there any practical reason to move away from the format `func (rule SimulationQuesta) Build(ctx core.Context)`?

Reply:
Yes. I find it very bad practice make things more complicated than necessary. There was no added functionality provided by the `SimulationQuesta` struct and it required needless copying of state. Using plain functions gives the exact same functionality with less code.